Feat/rangeSelector : RangeSelector 컴포넌트 구현

iyu88 opened the pull request:
## 🤠 개요

- #53 
- RangeSelector 컴포넌트를 구현합니다. 

## 💫 설명

- Compound Component 방식으로 만들어보고 싶어서 Context API를 사용했습니다. 처음 시도해보는데 꽤 많이 어려웠습니다. 🤯

- 일단 아직 부족한 부분이 많은데, 계속 붙잡고 있으면 다른 일 처리 못할까봐 RP 올려두고 차차 필요한 부분 덧붙이겠습니다. 

- 수많은 Ref가 존재하는데, DOM을 건드리는 작업 + 좌표를 고정적으로 저장하고 싶어서 그랬습니다. 
  뭔가 더 좋은 방법이 있을 것 같은데 알려주세요 🙏

- 접근성을 위해서 [MDN 의 Slider](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/slider_role) 를 참고하여 aria 속성을 부여했습니다. 

- useLayoutEffect 에서 Ref 관련 정보들을 먼저 다 받은 뒤에 해당 값들을 사용합니다. 근데 Ref.current 값이 없을 수도 있어서 조건문에서
  계속 값이 있는지 확인해주고 있습니다. Early return 하는 방식으로 바꿀까도 싶습니다. 

- useEffect 에서 등록하는 마우스 이벤트는 AbortController 를 사용해서 remove 하는 방식을 사용해보았습니다.

- RangeSelector 인데 현재는 Single Slider 라고 보시면 될 것 같습니다. 가능하다면 Props를 통해서 Double Thumb를 생성하도록 
  해보겠습니다. 개인적으로는 후술할 추가 작업 내용들 중에서 가장 우선순위가 낮다고 생각합니다. 
  <del>(만약 못만들면 이름을 RangeSlider 로 개명..)</del>

- Range (최대 최소 값의 차) 와 실제 Window 좌표를 비율로 계산해서 맞춰주려고 일반 함수들을 작성했습니다. 
  저에게는 너무 어려운 수학이었습니다. 지금은 호출할 때마다 모든 값을 다 전달하고 있는데 lint에 의해 가독성도 떨어지고 해서 Closer 를 
  이용하여 해결해 볼 계획입니다.

- Value 가 표시되는 부분에 대해서는 숫자가 커질수록 Thumb의 크기 때문에 문제가 될 수 있을 것 같은데 추후에 Tooltip 같은 컴포넌트를 
  만들게 된다면 Hover 했을 때에만 표시하는 방향을 염두에 두고 있습니다. 

- 키보드 이벤트로 조절할 수 있게 하려고 합니다. 이벤트 핸들러는 만들어놓았는데, 작성하다보니까 코드가 점점 더러워져서 Util 함수까지 
  구상하고나서 합쳐보겠습니다. 

- 지금 코드를 index.ts 파일 안에 모두 다 구현해놓았는데 😓 세세하게 분리하려다가 도중에 멈췄습니다. 이유는 Slider 컴포넌트 안에 있는 
  태그들을 분리할 때 적절한 컴포넌트 이름이 생각나지 않았기 때문입니다. 
  ex) Track 태그를 컴포넌트로 분리한다고 했을 때, 내부에 위치할 스타일 컴포넌트 이름은 무엇으로 해야하는지 등등

- 적다 보니까 이게 PR인지 이슈인지 참 모르겠지만~ 계속 조금씩 다듬어 보겠습니다. 귀한 리뷰 감사합니다. 🙇‍♂️


## 📷 스크린샷 (Optional)

#### 생성한 스토리 중에서 마지막 스토리 시연 영상입니다. 

https://user-images.githubusercontent.com/31645195/225994184-93f4382c-9689-421d-9559-ec407b94c50c.mov

iyu88 committed:
feat: RangeSelector UI 구현

iyu88 committed:
feat: Slider 이벤트 구현

iyu88 committed:
feat: Track 클릭 시 Thumb 이동 이벤트 핸들러 구현

iyu88 committed:
feat: Thumb 위치 설정 및 Slider 값 설정

iyu88 committed:
refactor: thumbWidth 변수 적용

iyu88 committed:
refactor: ref 변수 이름 수정

iyu88 committed:
refactor: FilledRef 를 Track::after로 대체

기존에 useRef 를 이용하여 FilledRef 를 사용하는 것이 불필요한 ref 를 추가한다고 생각했습니다.
Track 에  pseudo class 인 ::after 를 적용하여 Slider 가 채워진 효과를 줍니다.

이를 설정할 때 setProperty 메서드를 사용합니다.
동시에 Thumb 와 Track 에서 공통적으로 사용할 수 있도록 setSliderPos 함수를 정의했습니다.

iyu88 committed:
refactor: context 변수 destructuring

iyu88 committed:
fix: mousemove 이벤트가 중첩 등록되는 문제 해결

useEffect 안에서 이벤트를 등록하기 전에 movable 상태일 때만 등록하도록 조건을 걸어주어 중첩 등록을 막습니다.

iyu88 committed:
refactor: thumbWidth 를 offsetRef 로 변경

Slider에서 보장되어야 하는 간격을 나타내는 thumbWidth 값을 ref에 저장하여 LayoutEffect 에서 설정합니다.

iyu88 committed:
refactor: Slider 에서 사용하는 함수에서 offset 옵션을 받도록 변경

iyu88 committed:
feat: fixInt 함수 생성

매개변수를 소수점 첫번째 자리에서 반올림하여 반환하는 함수를 정의합니다.

iyu88 committed:
feat: Slide 색상 지정

외부에서 Color 를 주입받는 것이 아니라 내부에서 theme 으로 색상을 결정합니다.

iyu88 committed:
fix: calcRangeValue 곱셈 비율 수정

기존에는 계산된 값에 100을 곱함으로써 고정된 비율로 나누었던 점을 변동 비율로 나누도록 수정합니다.

iyu88 committed:
feat: 접근성 관련 속성 추가

RangeSelectorProps 에서 추가적으로 id, label 을 받아서 하위 컴포넌트에서 접근성 관련 속성을 명시합니다.

iyu88 committed:
fix: 최솟값이 0이 아닐 때 발생하는 문제 해결

최솟값이 0이 아닐 경우 track 의 filled 너비와 value 상태값이 비정상적으로 표시되는 문제를 해결합니다.
setSliderPos 에서 최종적으로 값을 곱할 때 현재 value 에서 최솟값을 뺀 값을 사용합니다.
상태를 업데이트할 때에도 min 값을 더해서 최솟값을 보장합니다.

iyu88 committed:
feat: RangeSelector 스토리 생성

prayinforrain commented on src/components/RangeSelector/index.tsx:
```diff
@@ -0,0 +1,266 @@
+    const controller = new AbortController();
+    const { signal } = controller;
+
+    if (movable) {
+      window.addEventListener(
+        'mousemove',
+        (e) => {
+          window.addEventListener('mouseup', () => {
+            setMovable(false);
+            controller.abort();
+          });
+
+          if (!movable) return;
+          if (e.clientX > maxPosRef.current) return;
+          if (e.clientX < minPosRef.current) return;
+          handleThumbPosition(e);
+        },
+        { signal },
+      );
+    }
```
이거 콘솔로그 찍어봤는데, move가 실행된 만큼 up 이벤트가 등록되고 제대로 abort 되지 않는 것 같아요.  
![image](https://user-images.githubusercontent.com/46566891/226080635-8b675e88-5be8-4aa3-bf7d-7d8dc17a7525.png)

mouseup 이벤트도 같은 signal 객체를 넣어주면 어떨까요?

iyu88 replied:
제가 놓쳤던 부분이네요! 수정 / 반영했습니다!

prayinforrain commented on src/components/RangeSelector/index.tsx:
```diff
@@ -0,0 +1,266 @@
+  max: number,
+  min: number,
+  value: number,
+  offset?: number,
+) => ((offset ? size - offset : size) / (max - min)) * (value - min);
```
삼항 연산자 말고 그냥 offset의 기본값을 0으로 지정해주면 어떨까요? 그게 더 보기 편할 것 같아요

iyu88 replied:
훨씬 더 깔끔한 것 같아요!! 👍

prayinforrain commented:
4시에 PR을 올리시면 어떡해요 ㅠㅠ.. 주무시라구요
고생하셨어요~~

보다가 궁금한게 생겼는데, 혹시 RangeSelector 외부에서 RangeSelector의 값을 참조할 때에는 어떻게 사용해야 하나요 ?ㅁ?  
form 내부에서 RangeSelector를 사용해서 제출할 때 값을 참조해야 하는 상황에서요!

iyu88 committed:
refactor: 이벤트 및 Slider 관련 함수 수정

mouseup 이벤트에도 이벤트 해제가 정상적으로 작동하도록 signal 객체를 전달합니다.
Slider의 위치와 표시되는 값을 계산하는 함수에 offset 의 기본값을 0으로 설정하여 가독성을 높입니다.

iyu88 commented:
@prayinforrain
아하 그러네요? 
이 부분은 Slider 내부에 안보이는 `<input type="range" />` 를 넣어두는 방식으로 해볼까 하는데, 
이렇게 하면 꺼내서 사용할 수 있겠죠...?!? 🤔

se030 commented on src/components/RangeSelector/index.tsx:
```diff
@@ -0,0 +1,267 @@
+    }
+  }, [movable]);
+
+  if (context === null) return null;
```
컴포넌트에서 null 리턴해도 괜찮은가요?

iyu88 replied:
`return <></>` 로 변경해볼게요

se030 commented on src/components/RangeSelector/index.tsx:
```diff
@@ -0,0 +1,267 @@
+  };
+
+  useLayoutEffect(() => {
+    if (trackRef.current && thumbRef.current && context) {
```
여기나 ln 174 같은거 early return으로 작성하는게 더 깔끔할 것 같아요!

iyu88 replied:
early return 반영했습니다. 확인 부탁드려요!

se030 commented on src/components/RangeSelector/index.tsx:
```diff
@@ -0,0 +1,267 @@
+  min,
+  max,
+  init,
+  size,
```
trackWidth로 어디에 사용되는 값인지 더 명확하게 나타내면 사용처에서 이해하기 더 좋을 것 같아요.

iyu88 replied:
trackWidth 으로 props 이름 변경 완료했어요~~

se030 reviewed (approved):
고생하셨습니다 !! 우재님이 기능은 코멘트 해주셔서 더 수정할게 없어보이네요.
개선 작업도 하신다면 별도 PR로 올려주세요.

근데 혹시 Compound 패턴 사용하신 이유가 있나요? 얘같은 경우에는 Slider나 RangeDisplay 같은 내부 컴포넌트가 외부에 인터페이스를 열어줄 필요가 없는 것 같아서요.

iyu88 committed:
refactor: Early return 적용 및 trackWidth 로 props 이름 수정

iyu88 committed:
refactor: interface 타입 분리

[ComponentName]ContextInterface 로 정한 인터페이스 네이밍 컨벤션을 적용합니다.

iyu88 committed:
refactor: RangeSelectorProps 를 사용한 인터페이스 선언
